tasks/update-granules-metadata/tests/test-index.js

```diff
@@ -0,0 +1,136 @@
+
+  newPayload.input.granules.forEach((gran) => {
+    gran.files.forEach((file) => {
+      file.bucket = t.context.stagingBucket;
+      file.filename = buildS3Uri(t.context.stagingBucket, parseS3Uri(file.filename).Key);
+    });
+  });
+
+  return newPayload;
+}
+
+test.beforeEach(async (t) => {
+  t.context.stagingBucket = randomId('staging');
+  t.context.publicBucket = randomId('public');
+  t.context.protectedBucket = randomId('protected');
+  t.context.systemBucket = randomId('system');
+  t.context.stackName = randomString();
+  await Promise.all([
+    s3().createBucket({ Bucket: t.context.stagingBucket }).promise(),
+    s3().createBucket({ Bucket: t.context.publicBucket }).promise(),
+    s3().createBucket({ Bucket: t.context.protectedBucket }).promise(),
+    s3().createBucket({ Bucket: t.context.systemBucket }).promise(),
+  ]);
+  process.env.system_bucket = t.context.systemBucket;
+  process.env.stackName = t.context.stackName;
+  putJsonS3Object(
+    t.context.systemBucket,
+    getDistributionBucketMapKey(t.context.stackName),
+    {
+      [t.context.stagingBucket]: t.context.stagingBucket,
+      [t.context.publicBucket]: t.context.publicBucket,
+      [t.context.protectedBucket]: t.context.protectedBucket,
+      [t.context.systemBucket]: t.context.systemBucket,
+    }
+  );
+
+  const payloadPath = path.join(__dirname, 'data', 'payload.json');
+  const rawPayload = fs.readFileSync(payloadPath, 'utf8');
+  t.context.payload = JSON.parse(rawPayload);
+  const filesToUpload = granulesToFileURIs(t.context.payload.input.granules);
+  t.context.filesToUpload = filesToUpload.map((file) =>
+    buildS3Uri(`${t.context.stagingBucket}`, parseS3Uri(file).Key));
+  process.env.REINGEST_GRANULE = false;
+});
+
+test.afterEach.always(async (t) => {
+  await recursivelyDeleteS3Bucket(t.context.publicBucket);
+  await recursivelyDeleteS3Bucket(t.context.stagingBucket);
+  await recursivelyDeleteS3Bucket(t.context.protectedBucket);
+  await recursivelyDeleteS3Bucket(t.context.systemBucket);
+});
+
+test.serial('Should add etag to each CMR metadata file by checking that etag is one of more characters not whitespace', async (t) => {
+  const newPayload = buildPayload(t);
+  const filesToUpload = cloneDeep(t.context.filesToUpload);
+  await uploadFiles(filesToUpload, t.context.stagingBucket);
+
+  const output = await updateGranulesMetadata(newPayload);
+
+  output.granules.forEach((g) => g.files
+    .filter(isCMRFile)
+    .forEach(({ etag = '' }) => t.regex(etag, /"\S+"/)));
+});
+
+test.serial('Should update existing etag on CMR metadata file', async (t) => {
+  const newPayload = buildPayload(t);
+  const filesToUpload = cloneDeep(t.context.filesToUpload);
+  await uploadFiles(filesToUpload, t.context.stagingBucket);
+  const inputGranule = newPayload.input.granules[newPayload.input.granules.findIndex((g) => g.granuleId === 'MOD11A1.A2017200.h19v04.006.2017201090722')];
```

Comment:
So the test is still depending on knowing the granuleId from the fixture. I think it could be refactored like this to avoid that dependency:

```
const inputGranulesWithEtags = newPayload.input.granules.reduce(
  (acc, granule) => {
    const cmrFile = granule.files.find(isCMRFile);
    if (!cmrFile) return acc;
    acc.push([granule.granuleId, cmrFile]);
  },
  []
);

inputGranulesWithEtags.forEach(([inputGranuleId, inputCmrFile]) => {
  const outputGranule = output.granules.find((outputGranule) => outputGranule.granuleId === inputGranuleId);
  if (outputGranule) {
    const outputCmrFile = outputGranule.files.find(isCMRFile);
    t.not(inputCmrFile.etag, outputCmrFile.etag)
  }
})
```

What do you think? Does that seem preferable or like unnecessary optimization? Also I didn't test the above code so it could need some tweaking

Reply:
Hm, I did it a little differently. https://github.com/nasa/cumulus/pull/1969/files#diff-41cb5ec44ef17d2ccd2ba9142b73b29aee1d0c497c17766f8877a8d58648aa57R112-R125

Let me know what you think!